libraries/deploy-manager/src/DeployManager.ts

```diff
@@ -0,0 +1,810 @@
+      if (sourceProjectConfiguration) {
+        useNpmIgnoreFilter = true;
+      }
+    }
+
+    const targetFolderPath: string = this._remapPathForDeployFolder(sourceFolderPath, options);
+
+    if (useNpmIgnoreFilter) {
+      // Use npm-packlist to filter the files.  Using the WalkerSync class (instead of the sync() API) ensures
+      // that "bundledDependencies" are not included.
+      const walker: npmPacklist.WalkerSync = new npmPacklist.WalkerSync({
+        path: sourceFolderPath
+      });
+      walker.start();
+      const npmPackFiles: string[] = walker.result;
+
+      const alreadyCopiedSourcePaths: Set<string> = new Set();
+
+      await Async.forEachAsync(
+        npmPackFiles,
+        async (npmPackFile: string) => {
+          // In issue https://github.com/microsoft/rushstack/issues/2121 we found that npm-packlist sometimes returns
+          // duplicate file paths, for example:
+          //
+          //   'dist//index.js'
+          //   'dist/index.js'
+          //
+          // We can detect the duplicates by comparing the path.resolve() result.
+          const copySourcePath: string = path.resolve(sourceFolderPath, npmPackFile);
+          if (alreadyCopiedSourcePaths.has(copySourcePath)) {
+            return;
+          }
+          alreadyCopiedSourcePaths.add(copySourcePath);
+
+          const copyDestinationPath: string = path.join(targetFolderPath, npmPackFile);
+
+          if (state.symlinkAnalyzer.analyzePath(copySourcePath).kind !== 'link') {
+            await FileSystem.ensureFolderAsync(path.dirname(copyDestinationPath));
+
+            await FileSystem.copyFileAsync({
+              sourcePath: copySourcePath,
+              destinationPath: copyDestinationPath,
+              alreadyExistsBehavior: AlreadyExistsBehavior.Error
+            });
+          }
+        },
+        {
+          concurrency: 10
+        }
+      );
+    } else {
+      // use a simplistic "ignore" ruleset to filter the files
+      const ignoreFilter: Ignore = ignore();
+      ignoreFilter.add([
+        // The top-level node_modules folder is always excluded
+        '/node_modules',
+        // Also exclude well-known folders that can contribute a lot of unnecessary files
+        '**/.git',
+        '**/.svn',
+        '**/.hg',
+        '**/.DS_Store'
+      ]);
+
+      await FileSystem.copyFilesAsync({
+        sourcePath: sourceFolderPath,
+        destinationPath: targetFolderPath,
+        alreadyExistsBehavior: AlreadyExistsBehavior.Error,
+        filter: async (src: string, dest: string) => {
+          const relativeSrc: string = path.relative(sourceFolderPath, src);
+          if (!relativeSrc) {
+            return true; // don't filter sourceFolderPath itself
+          }
+
+          if (ignoreFilter.ignores(relativeSrc)) {
+            return false;
+          }
+
+          const stats: FileSystemStats = FileSystem.getLinkStatistics(src);
```

Comment:
Leaving as-is. copyFilesAsync uses the underlying fsx.copy API, where the filter does not provide this information. It is available, but it would necessitate a change to fs-extra https://github.com/jprichardson/node-fs-extra/blob/cc7b3b22a984de5124131c7897574091c9df00e1/lib/copy/copy.js#L38